### ISSUE TITLE
mpg123 does not return @P 3 at the end of a song

### ISSUE THREAD

**Issue by stefets** [4br3mm0rd/mpyg321#15]
Hey Emmanuel!

mpg123 does not return @ P 3 at the end of a song but it returns @ P 0; i have a recent version of mpg123 --version = 1.27.2

This break the mpg_outs array logic, there is a conflict between "user_stop" and "end_of_song", actualy in mpg123 the response is @ P 0 in both cases

Maybe @ P 3 is a mpg321 behaviour ? I don't use it, can you double check please ?

I want to discuss with you for a solution; actualy, your library is hybrid for mpg123 and mpg321; and there's no abstraction between them except in the constructor; mpg321 is an old program, not updated since 2012 and mpg123 is realy up to date; can you please consider the obsolescence of the support of mpg321 in your library ? But it's not a big deal to support both version with their differences.

Let me know what you think!

Merci !

--

This is my output test, if i send 'stop', it returns @ P 0, if I let it play all, it returns @ P 0
```
stefets@rpi3:/media/soundlib/mp3 $ mpg123 --remote
@R MPG123 (ThOr) v9
silence
@silence
load gong.mp3
@I gong
@P 2
@S 1.0 3 48000 Joint-Stereo 0 960 2 0 0 0 320 0 0
stop
@P 0
load gong.mp3
@I gong
@P 2
@S 1.0 3 48000 Joint-Stereo 0 960 2 0 0 0 320 0 0
@P 0
```

**Comment by 4br3mm0rd** [comment]
As I said in PR #14 , do you think we can detect which program is running and adapt to it?
Ideally not by the name, although it is a viable solution. Definitely not my by signature :) 
Maybe by parsing the -h command? Or any command that might give us info, with a default value set to mpg123 (just because the other one is not being updated - and because at some point, if your environment is too funky, then you should adapt a bit to the wrapper).

What do you think?

**Comment by stefets** [comment]
Hello! yes we have to know the player version context to keep it hybrid.

I suggest a class

```
class PlayerVersion:
  MPG123 = 0
  MPG321 = 1
```

In the constructor if mpg321 does not failed, we set a property self.PlayerVersion = PlayerVersion.MPG321, and same logic for mpg123. Then in the code we can check the PlayerVersion

**Comment by 4br3mm0rd** [comment]
OK, this solves the first part of our issue. I will try to implement this by the end of the week.

About the second part, interestingly enough, I don't see "@P 3" in the documentation of mpg321 ([https://github.com/vborrego/mpg321/blob/master/README.remote](see here)). I probably saw it during my tests. 

We can either delete the "user_stop" feature, or we can keep it and have a specific behavior dealing with it:
- @P 0 would be a "stop event"
- When the user calls "stop()", we set a property in the object stating that the user is stopping the song
- When the stop event is called, we check this property
I think this is robust enough, because no one but the program is supposed to interact with mpg. What do you think?

**Comment by stefets** [comment]
Hi,!

I ike your idea for the P0, fit with mine ! We have to know when stop is called. 

P3 is in the mad.c source code and I don't know what it means.

**Comment by 4br3mm0rd** [comment]
OK I'll implement that :)

**Comment by stefets** [comment]
Thanks!
This is essential to manage the particalarities correctly.
I think to reject my PR and start a new branch after those fixes.

**Comment by 4br3mm0rd** [comment]
Yeah, let's do that and separate all the features to make it clean (so that if there is a review on one of them, it won't block the development of other features :-) )

**Comment by stefets** [comment]
Perfect!

I think giving the full control to the programmer by passing an object to the constructor can be a good addition in the future, instead of 'guessing' the player

Pseudo:

player_configuration = PlayerConfiguration()
player_configuration.version = PlayerVersion.MPG123
player_configuration.add_args('audiodevice', 'hw:1,0')
player_configuration.add_args(key, value')
player_configuration.volume_min = 20
player_configuration.volume_max = 80
player_configuration.futur = None

self.player = MPyg321Player(player_configuration)

What do you think ?

**Comment by 4br3mm0rd** [comment]
I think it would be more python-ish to have parameters in the constructor and to create the configuration object inside the constructor.
You would then call the constructor with: 
```
MPyg321Player(version=PlayerVersion.MPG123, audiodevice="hw:1,0")
```
This also allows the user not to care about the object.

**Comment by stefets** [comment]
Alright! Then we will see if an object is necessary in the library, maybe not. Thanks!

**Comment by 4br3mm0rd** [comment]
According to the subject of this issue, I just made a test, and mpg321 does not return any `@P 0` at the end of the song, but only a `@P 3`. So we need to define different behaviors according to the version :-)

**Comment by stefets** [comment]
Hi  @4br3mm0rd !

Hope you're doing well ? 

FYI, the latest release of mpg123 is now 1.30 and contains our @ P3 feature request.

https://www.mpg123.de/#2022-06-26

**Comment by 4br3mm0rd** [comment]
Hey @stefets 

Thanks for the update. From what I remember we decided not to use it in our implementation in the end right?

**Comment by stefets** [comment]
Indeed! 

But I have a use-case and the MPyg123Player lack the feature and it's where the P3 parameter is needed for me. 

I will add the logic in my child class, override the load_list maybe. I want to be able to play an entire playlist, that's it.

And when P3 arrise, the player will play next song on the playlist and so on until it reach the end or restart to the begining for instance.

**Comment by 4br3mm0rd** [comment]
Can't you just implement this logic inside the internal "end_of_song" callback?

**Comment by stefets** [comment]
Yes surely for the callback! 

I will inject the P3 code and the logic in my child class, I dont wan't to fork the wrapper at this moment but I think the P3 code could be part of the array defined in the const.py file. It is the responsibility of the programmer who inherits from the base class to manage the P3 code and if he does not have the correct installation of the mpg123 program version 1.30 and + on his system he must be aware that it is impossible to implement it in any way, and this can be commented in the code.

Anyway! It will work on my side and my implementation will be, i think, a good example!

**Comment by 4br3mm0rd** [comment]
I'm not sure I understand the use of what you are trying to do but if you think it is relevant, feel free to add the example to the examples folder in the project :)

**Comment by stefets** [comment]
Hey @4br3mm0rd, 

I have made the changes in my child class to take care of the P3 code available in the 1.30 version of mpg123. The implementation was flawlessly easy. My goal was to play a playlist automaticaly; The P3 code is the life saver.

I don't think adding my implementation in your repo example directory is the way to go. The wrapper is manipulated through MIDI messages and MIDIDINGS and it's an overkill example; However, I suggest to add my repo as a link in your WIKI, I think it's a good place for references and examples.

I have merged a working implementation in my main branch
https://github.com/stefets/live-config/blob/main/src/plugins/audioplayer/mp3.py

Thanks!
